vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc

```diff
@@ -743,6 +859,61 @@ bool AdsImpl::TestSearchState(
   return is_search_engine;
 }
 
+void AdsImpl::TestCheckoutStateAndAdHistory(const std::string& url,
+    std::map<std::string, std::string> campaign_checkout_urls) {
+  // Iterate over checkout URLs
+  for (auto elem : campaign_checkout_urls) {
+    std::string campaign_id = elem.first;
+    std::string checkout_url = elem.second;
+
+    // If checkout URL matches and campaign found in ad history
+    // consider test successful
+    if (url == checkout_url) {
+      BLOG(INFO) << "Client in checkout state at " << url;
+      auto campaign_history = client_->GetCampaignHistory();
+
+      if (campaign_history.find(campaign_id) !=
+          campaign_history.end()) {
+        BLOG(INFO) << "Attribute conversion to campaign " << campaign_id;
+      }
+    }
+  }
+}
+
+void AdsImpl::ServeSegmentAdIfPurchaseIntent() {
+  // TODO(Moritz Haller): Have checks from AdsImpl::CheckAdReady
+
+  auto segment_history = client_->GetSegmentHistory();
+
+  size_t total_oci_signal_count = 0;
+  size_t max_oci_signals = 0;
+  std::string winning_segment;
+
+  for (auto x : segment_history) {
+      std::string segment = x.first;
+      size_t oci_signals = x.second.size();
+
+      if (oci_signals > max_oci_signals) {
+          max_oci_signals = oci_signals;
+          winning_segment = segment;
+      }
+
+      total_oci_signal_count += oci_signals;
+  }
+
+  BLOG(INFO) << "OCI Signals received: " << total_oci_signal_count
+    << "; Winning segment: " << winning_segment << std::endl;
+
+  // Consider n actions in either shopping or search state as having
+  // purchase intent
+  size_t n_actions = 3;  // TODO(Moritz Haller): move to static_values.h
+  if (total_oci_signal_count >= n_actions) {
+    BLOG(INFO) << "Serve ad for segment " << winning_segment << " is ready";
+    ServeAdFromCategory(winning_segment);
```

Comment:
We should not be serving the ad from here as business logic such as if the app is foreground etc. should be taken into account, please see above comment.